ProjFS.Mac/PrjFSKext/VnodeCache.cpp

```diff
@@ -0,0 +1,381 @@
+    const FsidInode& vnodeFsidInode,
+    uintptr_t startingIndex,
+    uint32_t vnodeVid,
+    /* out paramaeters */
+    VirtualizationRootHandle& rootHandle);
+
+struct VnodeCacheEntry
+{
+    vnode_t vnode;
+    uint32_t vid;   // vnode generation number
+    VirtualizationRootHandle virtualizationRoot;
+};
+
+KEXT_STATIC uint32_t s_entriesCapacity;
+KEXT_STATIC VnodeCacheEntry* s_entries;
+static RWLock s_entriesLock;
+static const uint32_t MinEntriesCapacity = 0x040000; //  4 MB (assuming 16 bytes per VnodeCacheEntry)
+static const uint32_t MaxEntriesCapacity = 0x400000; // 64 MB (assuming 16 bytes per VnodeCacheEntry)
+
+kern_return_t VnodeCache_Init()
+{
+    if (RWLock_IsValid(s_entriesLock))
+    {
+        return KERN_FAILURE;
+    }
+    
+    s_entriesLock = RWLock_Alloc();
+    if (!RWLock_IsValid(s_entriesLock))
+    {
+        return KERN_FAILURE;
+    }
+
+    s_entriesCapacity = MAX(MinEntriesCapacity, desiredvnodes * 2);
+    s_entriesCapacity = MIN(MaxEntriesCapacity, s_entriesCapacity);
+    
+    s_entries = Memory_AllocArray<VnodeCacheEntry>(s_entriesCapacity);
+    if (nullptr == s_entries)
+    {
+        s_entriesCapacity = 0;
+        return KERN_RESOURCE_SHORTAGE;
+    }
+    
+    VnodeCache_InvalidateCache(nullptr);
+    
+    PerfTracing_RecordSample(PrjFSPerfCounter_CacheCapacity, 0, s_entriesCapacity);
+    
+    return KERN_SUCCESS;
+}
+
+void VnodeCache_Cleanup()
+{
+    if (nullptr != s_entries)
+    {
+        Memory_FreeArray<VnodeCacheEntry>(s_entries, s_entriesCapacity);
+        s_entries = nullptr;
+        s_entriesCapacity = 0;
+    }
+    
+    if (RWLock_IsValid(s_entriesLock))
+    {
+        RWLock_FreeMemory(&s_entriesLock);
+    }
+}
+
+VirtualizationRootHandle VnodeCache_FindRootForVnode(
+    PerfTracer* _Nonnull perfTracer,
+    PrjFSPerfCounter cacheHitCounter,
+    PrjFSPerfCounter cacheMissCounter,
+    PrjFSPerfCounter cacheMissFallbackFunctionCounter,
+    PrjFSPerfCounter cacheMissFallbackFunctionInnerLoopCounter,
+    vnode_t _Nonnull vnode,
+    const FsidInode& vnodeFsidInode,
+    bool invalidateEntry)
+{
+    bool lockElevatedToExclusive = false;
+    VirtualizationRootHandle rootHandle = RootHandle_None;
+    uintptr_t indexFromHash = HashVnode(vnode);
+    uint32_t vnodeVid = vnode_vid(vnode);
+    
+    RWLock_AcquireShared(s_entriesLock);
+    {
+        uintptr_t cacheIndex;
+        if (TryFindVnodeIndex_SharedLocked(vnode, indexFromHash, indexFromHash, /*out*/ cacheIndex))
+        {
+            if (vnode == s_entries[cacheIndex].vnode)
+            {
+                if (invalidateEntry || vnodeVid != s_entries[cacheIndex].vid)
+                {
+                    perfTracer->IncrementCount(cacheMissCounter, true /*ignoreSampling*/);
+                
+                    UpgradeToExclusiveLock(s_entriesLock);
+                    lockElevatedToExclusive = true;
+                    
+                    UpdateCacheEntryToLatest_ExclusiveLocked(
```

Comment:
I don't think this call is safe in the case where `UpgradeToExclusiveLock` failed internally. If the cache was invalidated by another thread in the time between when the lock was dropped and reacquired, this will insert an item at location `cacheIndex`, which may no longer be the correct place for it if there was previously a collision with the item at `indexFromHash`, which there may no longer be after invalidating the cache.

Reply:
I was thinking the same here.